numpyro/contrib/hsgp/laplacian.py

```diff
@@ -51,38 +128,76 @@ def eigenfunctions(x: ArrayImpl, ell: float, m: int) -> ArrayImpl:
-    Stat Comput 30, 419-446 (2020)
+
+        1. Solin, A., Särkkä, S. Hilbert space methods for reduced-rank Gaussian process regression.
+           Stat Comput 30, 419-446 (2020)
 
     :param ArrayImpl x: The points at which to evaluate the eigenfunctions.
-    :param float ell: The length of the interval divided by 2.
-    :param int m: The number of eigenfunctions to compute.
-    :returns: An array of the first `m` eigenfunctions evaluated at `x`.
+        If `x` is 1D the problem is assumed unidimensional.
+        Otherwise, the dimension of the input space is inferred as the last dimension of `x`.
+        Other dimensions are treated as batch dimensions.
+    :param float | list[float] ell: The length of the interval in each dimension divided by 2.
+        If a float, the same length is used in each dimension.
+    :param int | list[int] m: The number of eigenvalues to compute in each dimension.
+        If an integer, the same number of eigenvalues is computed in each dimension.
+    :returns: An array of the first :math:`m^\\star \\times D` eigenfunctions evaluated at `x`.
     :rtype: ArrayImpl
     """
-    m1 = (jnp.pi / (2 * ell)) * jnp.tile(ell + x[:, None], m)
-    m2 = jnp.diag(jnp.linspace(1, m, num=m))
-    num = jnp.sin(m1 @ m2)
-    den = jnp.sqrt(ell)
-    return num / den
-
-
-def eigenfunctions_periodic(x, w0, m):
+    if x.ndim == 1:
+        x_ = x[..., None]
+    else:
+        x_ = x
+    dim = x_.shape[-1]  # others assumed batch dims
+    n_batch_dims = x_.ndim - 1
+    ell_ = _convert_ell(ell, dim)
+    a = jnp.expand_dims(ell_, tuple(range(n_batch_dims)))
+    b = jnp.expand_dims(sqrt_eigenvalues(ell_, m, dim), tuple(range(n_batch_dims)))
+    return jnp.prod(jnp.sqrt(1 / a) * jnp.sin(b * (x_[..., None] + a)), axis=-2)
+
+
+def eigenfunctions_periodic(x: ArrayImpl, w0: float, m: int):
     """
     Basis functions for the approximation of the periodic kernel.
 
-    :param x: The points at which to evaluate the eigenfunctions.
-    :param w0: The frequency of the periodic kernel.
-    :param m: The number of eigenfunctions to compute.
+    :param ArrayImpl x: The points at which to evaluate the eigenfunctions.
+    :param float w0: The frequency of the periodic kernel.
+    :param int m: The number of eigenfunctions to compute.
 
     .. note::
         If you want to parameterize it with respect to the period use `w0 = 2 * jnp.pi / period`.
+
+    .. warning::
+        Multidimensional inputs are not supported.
     """
+    if x.ndim > 1:
+        raise ValueError(
+            "Multidimensional inputs are not supported by the periodic kernel."
+        )
     m1 = jnp.tile(w0 * x[:, None], m)
     m2 = jnp.diag(jnp.arange(m, dtype=jnp.float32))
     mw0x = m1 @ m2
     cosines = jnp.cos(mw0x)
     sines = jnp.sin(mw0x)
     return cosines, sines
+
+
+def _convert_ell(ell: float | list[float] | ArrayImpl, dim: int) -> ArrayImpl:
```

Comment:
Could we add a unit-test for this function? It would be great for documenting how is supposed to work.

Reply:
Added!